python/ray/data/tests/test_streaming_executor.py

```diff
@@ -692,6 +692,67 @@ def test_execution_allowed_nothrottle():
     )
 
 
+@pytest.mark.parametrize(
+    "max_errored_blocks, num_errored_blocks",
+    [
+        (0, 0),
+        (0, 1),
+        (2, 1),
+        (2, 2),
+        (2, 3),
+        (-1, 5),
+    ],
+)
+def test_max_errored_blocks(
+    ray_start_regular_shared,
+    restore_data_context,
+    max_errored_blocks,
+    num_errored_blocks,
+):
+    """Test DataContext.max_errored_blocks."""
+    num_tasks = 5
+
+    ctx = ray.data.DataContext.get_current()
+    ctx.max_errored_blocks = max_errored_blocks
+
+    def map_func(row):
+        id = row["id"]
+        if id < num_errored_blocks:
+            # Fail the first num_errored_tasks tasks.
+            raise RuntimeError(f"Task failed: {id}")
+        return row
+
+    ds = ray.data.range(num_tasks, parallelism=num_tasks).map(map_func)
+    should_fail = 0 <= max_errored_blocks < num_errored_blocks
+    if should_fail:
+        with pytest.raises(Exception, match="Task failed"):
+            res = ds.take_all()
+    else:
+        res = sorted([row["id"] for row in ds.take_all()])
+        assert res == list(range(num_errored_blocks, num_tasks))
+        stats = ds._get_stats_summary()
+        assert stats.extra_metrics["num_tasks_failed"] == num_errored_blocks
+
+
+def test_exception_concise_stacktrace(ray_start_regular_shared):
+    driver_script = """
+import ray
+
+def map(_):
+    raise ValueError("foo")
+
+ray.data.range(1).map(map).take_all()
+    """
+    proc = run_string_as_driver_nonblocking(driver_script)
+    out_str = proc.stdout.read().decode("utf-8") + proc.stderr.read().decode("utf-8")
+    # Test that the stack trace only contains the UDF exception, but not any other
+    # exceptions raised when the executor is handling the UDF exception.
+    assert (
+        "During handling of the above exception, another exception occurred"
+        not in out_str
+    ), out_str
```

Comment:
This PR also makes the error stack trace more concise. cc @stephanie-wang @c21 

Previously it will print:
```
Traceback (most recent call last):
  File "python/ray/_raylet.pyx", line 347, in ray._raylet.StreamingObjectRefGenerator._next_sync
  File "python/ray/_raylet.pyx", line 4653, in ray._raylet.CoreWorker.try_read_next_object_ref_stream
  File "python/ray/_raylet.pyx", line 447, in ray._raylet.check_status
ray.exceptions.ObjectRefStreamEndOfStreamError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/interfaces/physical_operator.py", line 80, in on_data_ready
    meta = ray.get(next(self._streaming_gen))
  File "python/ray/_raylet.pyx", line 302, in ray._raylet.StreamingObjectRefGenerator.__next__
  File "python/ray/_raylet.pyx", line 365, in ray._raylet.StreamingObjectRefGenerator._next_sync
StopIteration

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<string>", line 8, in <module>
  File "/Users/chenh/code/ray/python/ray/data/dataset.py", line 2459, in take_all
    for row in self.iter_rows():
  File "/Users/chenh/code/ray/python/ray/data/iterator.py", line 225, in _wrapped_iterator
    for batch in batch_iterable:
  File "/Users/chenh/code/ray/python/ray/data/iterator.py", line 183, in _create_iterator
    for batch in iterator:
  File "/Users/chenh/code/ray/python/ray/data/_internal/block_batching/iter_batches.py", line 176, in iter_batches
    next_batch = next(async_batch_iter)
  File "/Users/chenh/code/ray/python/ray/data/_internal/util.py", line 851, in make_async_gen
    raise next_item
  File "/Users/chenh/code/ray/python/ray/data/_internal/util.py", line 828, in execute_computation
    for item in fn(thread_safe_generator):
  File "/Users/chenh/code/ray/python/ray/data/_internal/block_batching/iter_batches.py", line 167, in _async_iter_batches
    yield from extract_data_from_batch(batch_iter)
  File "/Users/chenh/code/ray/python/ray/data/_internal/block_batching/util.py", line 210, in extract_data_from_batch
    for batch in batch_iter:
  File "/Users/chenh/code/ray/python/ray/data/_internal/block_batching/iter_batches.py", line 306, in restore_original_order
    for batch in batch_iter:
  File "/Users/chenh/code/ray/python/ray/data/_internal/block_batching/iter_batches.py", line 218, in threadpool_computations_format_collate
    yield from formatted_batch_iter
  File "/Users/chenh/code/ray/python/ray/data/_internal/block_batching/util.py", line 158, in format_batches
    for batch in block_iter:
  File "/Users/chenh/code/ray/python/ray/data/_internal/block_batching/util.py", line 117, in blocks_to_batches
    for block in block_iter:
  File "/Users/chenh/code/ray/python/ray/data/_internal/block_batching/util.py", line 54, in resolve_block_refs
    for block_ref in block_ref_iter:
  File "/Users/chenh/code/ray/python/ray/data/_internal/block_batching/iter_batches.py", line 254, in prefetch_batches_locally
    for block_ref, metadata in block_ref_iter:
  File "/Users/chenh/code/ray/python/ray/data/_internal/util.py", line 808, in __next__
    return next(self.it)
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/legacy_compat.py", line 54, in execute_to_legacy_block_iterator
    for bundle in bundle_iter:
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/interfaces/executor.py", line 37, in __next__
    return self.get_next()
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/streaming_executor.py", line 154, in get_next
    raise item
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/streaming_executor.py", line 223, in run
    while self._scheduling_loop_step(self._topology) and not self._shutdown:
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/streaming_executor.py", line 271, in _scheduling_loop_step
    num_errored_blocks = process_completed_tasks(
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/streaming_executor_state.py", line 416, in process_completed_tasks
    raise e
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/streaming_executor_state.py", line 383, in process_completed_tasks
    num_blocks_read = task.on_data_ready(
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/interfaces/physical_operator.py", line 93, in on_data_ready
    raise ex
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/interfaces/physical_operator.py", line 89, in on_data_ready
    ray.get(block_ref)
  File "/Users/chenh/code/ray/python/ray/_private/auto_init_hook.py", line 24, in auto_init_wrapper
    return fn(*args, **kwargs)
  File "/Users/chenh/code/ray/python/ray/_private/client_mode_hook.py", line 103, in wrapper
    return func(*args, **kwargs)
  File "/Users/chenh/code/ray/python/ray/_private/worker.py", line 2595, in get
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(ValueError): ray::Map(map)() (pid=94579, ip=127.0.0.1)
    for b_out in map_transformer.apply_transform(iter(blocks), ctx):
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/operators/map_transformer.py", line 371, in __call__
    for data in iter:
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/operators/map_transformer.py", line 196, in __call__
    yield from self._row_fn(input, ctx)
  File "/Users/chenh/code/ray/python/ray/data/_internal/planner/plan_udf_map_op.py", line 234, in transform_fn
    out_row = fn(row)
  File "/Users/chenh/code/ray/python/ray/data/_internal/planner/plan_udf_map_op.py", line 120, in fn
    return op_fn(item, *fn_args, **fn_kwargs)
  File "<string>", line 6, in map
ValueError: foo
```
Now it prints
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<string>", line 8, in <module>
  File "/Users/chenh/code/ray/python/ray/data/dataset.py", line 2459, in take_all
    for row in self.iter_rows():
  File "/Users/chenh/code/ray/python/ray/data/iterator.py", line 225, in _wrapped_iterator
    for batch in batch_iterable:
  File "/Users/chenh/code/ray/python/ray/data/iterator.py", line 183, in _create_iterator
    for batch in iterator:
  File "/Users/chenh/code/ray/python/ray/data/_internal/block_batching/iter_batches.py", line 176, in iter_batches
    next_batch = next(async_batch_iter)
  File "/Users/chenh/code/ray/python/ray/data/_internal/util.py", line 851, in make_async_gen
    raise next_item
  File "/Users/chenh/code/ray/python/ray/data/_internal/util.py", line 828, in execute_computation
    for item in fn(thread_safe_generator):
  File "/Users/chenh/code/ray/python/ray/data/_internal/block_batching/iter_batches.py", line 167, in _async_iter_batches
    yield from extract_data_from_batch(batch_iter)
  File "/Users/chenh/code/ray/python/ray/data/_internal/block_batching/util.py", line 210, in extract_data_from_batch
    for batch in batch_iter:
  File "/Users/chenh/code/ray/python/ray/data/_internal/block_batching/iter_batches.py", line 306, in restore_original_order
    for batch in batch_iter:
  File "/Users/chenh/code/ray/python/ray/data/_internal/block_batching/iter_batches.py", line 218, in threadpool_computations_format_collate
    yield from formatted_batch_iter
  File "/Users/chenh/code/ray/python/ray/data/_internal/block_batching/util.py", line 158, in format_batches
    for batch in block_iter:
  File "/Users/chenh/code/ray/python/ray/data/_internal/block_batching/util.py", line 117, in blocks_to_batches
    for block in block_iter:
  File "/Users/chenh/code/ray/python/ray/data/_internal/block_batching/util.py", line 54, in resolve_block_refs
    for block_ref in block_ref_iter:
  File "/Users/chenh/code/ray/python/ray/data/_internal/block_batching/iter_batches.py", line 254, in prefetch_batches_locally
    for block_ref, metadata in block_ref_iter:
  File "/Users/chenh/code/ray/python/ray/data/_internal/util.py", line 808, in __next__
    return next(self.it)
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/legacy_compat.py", line 54, in execute_to_legacy_block_iterator
    for bundle in bundle_iter:
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/interfaces/executor.py", line 37, in __next__
    return self.get_next()
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/streaming_executor.py", line 154, in get_next
    raise item
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/streaming_executor.py", line 223, in run
    while self._scheduling_loop_step(self._topology) and not self._shutdown:
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/streaming_executor.py", line 271, in _scheduling_loop_step
    num_errored_blocks = process_completed_tasks(
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/streaming_executor_state.py", line 416, in process_completed_tasks
    raise e from None
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/streaming_executor_state.py", line 383, in process_completed_tasks
    num_blocks_read = task.on_data_ready(
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/interfaces/physical_operator.py", line 93, in on_data_ready
    raise ex from None
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/interfaces/physical_operator.py", line 89, in on_data_ready
    ray.get(block_ref)
  File "/Users/chenh/code/ray/python/ray/_private/auto_init_hook.py", line 24, in auto_init_wrapper
    return fn(*args, **kwargs)
  File "/Users/chenh/code/ray/python/ray/_private/client_mode_hook.py", line 103, in wrapper
    return func(*args, **kwargs)
  File "/Users/chenh/code/ray/python/ray/_private/worker.py", line 2595, in get
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(ValueError): ray::Map(map)() (pid=95483, ip=127.0.0.1)
    for b_out in map_transformer.apply_transform(iter(blocks), ctx):
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/operators/map_transformer.py", line 371, in __call__
    for data in iter:
  File "/Users/chenh/code/ray/python/ray/data/_internal/execution/operators/map_transformer.py", line 196, in __call__
    yield from self._row_fn(input, ctx)
  File "/Users/chenh/code/ray/python/ray/data/_internal/planner/plan_udf_map_op.py", line 234, in transform_fn
    out_row = fn(row)
  File "/Users/chenh/code/ray/python/ray/data/_internal/planner/plan_udf_map_op.py", line 120, in fn
    return op_fn(item, *fn_args, **fn_kwargs)
  File "<string>", line 6, in map
ValueError: foo
```